build-and-deploy-docs/create_mkdocs_config.py

```diff
@@ -0,0 +1,342 @@
+#!/usr/bin/env python3
+"""
+Create MKDocs config yaml.
+
+If a README file is given, the content is split into individual markdown file
+for MKDocs to render.
+"""
+import argparse
+import collections
+import itertools
+import os
+import re
+import shutil
+
+from urllib.parse import urlparse, urlunparse
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import Dict, List
+
+import magic
+import mdformat
+import yaml
+
+from markdown_it import MarkdownIt
+from markdown_it.token import Token
+
+
+VALID_IMAGE_MIME_TYPES = {
+    'image/png',
+    'image/jpeg',
+    'image/pjpeg'
+    'image/gif',
+    'image/tiff',
+    'image/x-tiff',
+    'image/svg+xml'
+}
+
+
+def parse_args():
+    """ parse args """
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '--pipeline-dir',
+        type=Path,
+        required=True,
+        help='Path to the pipeline directory. Should be set to '
+        'GITHUB_WORKSPACE when called from github action.'
+    )
+    parser.add_argument(
+        '--pipeline-repo',
+        type=str,
+        required=True,
+        help='Pipeline repo name. Should be set to GITHUB_REPOSITORY '
+        ' when called from github action.'
+    )
+    parser.add_argument(
+        '--mkdocs-config',
+        type=Path,
+        help='Additional MKDocs config file.',
+        default=None
+    )
+    parser.add_argument(
+        '--readme',
+        type=Path,
+        help='Relative path to the README.md file.',
+        default=Path('README.md')
+    )
+
+    args = parser.parse_args()
+
+    # Handle GitHub Actions passing the literal default value "None"
+    if args.mkdocs_config is not None and args.mkdocs_config.name == 'None':
+        args.mkdocs_config = None
+
+    # Make sure the referenced files exist and are within this repository
+    if args.mkdocs_config is not None:
+        args.mkdocs_config = (args.pipeline_dir / args.mkdocs_config).resolve()
+        if not args.mkdoc_config.exists():
+            parser.error(f"Config file {args.mkdocs_config} not found!")
+
+        if not args.mkdocs_config.is_relative_to(args.pipeline_dir):
```

Comment:
Consider moving this above `.exists()`. It is possible to check the existence of files outside of the repository by inputting something like
```python
$ python create_mkdocs_config.py --mkdocs-config="etc/passwd" --pipeline-dir="../../../../.." --pipeline-repo='.'

create_mkdocs_config.py: error: Config file /etc/passwd outside of repository!
```
This can be used to discover otherwise hidden information about a system.

Reply:
Good catch!